docs/src/main/sphinx/develop/client-protocol.rst

```diff
@@ -0,0 +1,257 @@
+======================
+Trino client REST API
+======================
+
+The Trino client allows users to submit Trino queries and view results.  This
+note documents the REST API used by the Trino client.
+
+HTTP methods
+============
+
+* A ``POST`` to ``/v1/statement`` runs the query string in the ``POST`` body,
+  and returns a JSON document containing the query results.  If there are more
+  results, the JSON document will contain a ``nextUri`` URL attribute.
+* A ``GET`` to the ``nextUri`` attribute returns the next batch of query results.
+* A ``DELETE`` to ``nextUri`` terminates a running query.
+
+Overview of query processing
+============================
+
+A Trino client request is initiated by an HTTP ``POST`` to the endpoint
+``/v1/statement``, with a ``POST`` body consisting of the SQL query string.
+The caller may set header ``X-Trino-User`` to the username for the session,
+as well as a long list of other headers, documented below.
+
+If the client request returns an HTTP 503, that means the server was busy,
+and the client should try again in 50-100 milliseconds.  Any HTTP status other
+than 503 or 200 means that the query has failed.
+
+The ``/v1/statement`` ``POST`` request returns a JSON document of type
+``QueryResults``, as well as a collection of response headers.  The
+``QueryResults`` document will contain an ``error`` field of type
+``QueryError`` if the query has failed, and if that object is not present,
+the query succeeded.  Important members of ``QueryResults`` are documented
```

Comment:
are documented in the following section.

Reply:
Changed.